torch/csrc/jit/codegen/cuda/root_domain_map.cpp

```diff
@@ -0,0 +1,379 @@
+  const auto in_root = TensorDomain::noReductions(in_td->getRootDomain());
+  const auto& out_root = out_td->getRootDomain();
+  const auto& bcast_dim_flags = op->getBroadcastDimFlags();
+  TORCH_INTERNAL_ASSERT(out_root.size() == bcast_dim_flags.size(),
+                        "dim flags: ", bcast_dim_flags,
+                        ", out root: ", out_root);
+  auto in_it = in_root.begin();
+  auto out_it = out_root.begin();
+  while (in_it != in_root.end() && out_it != out_root.end()) {
+    if (bcast_dim_flags.at(std::distance(out_root.begin(), out_it))) {
+      // new broadcast dim. No matching dimension in the input
+      // tensor.
+      ++out_it;
+      continue;
+    }
+    attemptToProveId(in_td, *in_it, out_td, *out_it);
+    ++in_it;
+    ++out_it;
+  }
+  // At this point, the input domain should have been scanned
+  // entirely.
+  TORCH_INTERNAL_ASSERT(in_it == in_root.end(),
+                        "Unmatched domain detected: ", *in_it, " of ", in_td);
+  // On the other hand, the output may still have some domains left,
+  // and they must be new broadcast domains.
+  for (; out_it != out_root.end(); ++out_it) {
+    TORCH_INTERNAL_ASSERT(bcast_dim_flags.at(std::distance(out_root.begin(), out_it)),
+                          "Unmatched domain detected: ", *out_it, " of ", out_td);
+  }
+}
+
+bool RootDomainMap::mapAllConsumers(const DomainKey& producer_key) {
+  //std::cerr << "mapAllConsumers for : " << producer_key << std::endl;
+  auto it = pending_map_.find(producer_key);
+  TORCH_INTERNAL_ASSERT(it != pending_map_.end());
+  const auto& consumer_set = it->second;
+  // All entries in key_set must be equivalent with each other.
+  TORCH_INTERNAL_ASSERT(consumer_set.size() > 0);
+  bool consistent = safeToJoin(consumer_set, eq_set_, incompatible_domains_);
+  if (consistent) {
+    for (const auto pending_consumer: consumer_set) {
+#if 0
+      std::cerr << "Equivalent Ids found: " << producer_key << " == " << pending_consumer
+                << std::endl;
+#endif
+      proveId(producer_key, pending_consumer);
+    }
+  }
+  // This entry should never be used again, so remove it.
+  pending_map_.erase(it);
+  return consistent;
+}
+
+void RootDomainMap::handle(TensorView* tv) {
+  if (false) {
+    std::stringstream ss;
+    ss << tv;
+    std::cerr << "Visiting TensorView: " << ss.str() << std::endl;
+  }
+  const TensorDomain* td = tv->domain();
+  const auto root = TensorDomain::noReductions(td->getMaybeRFactorDomain());
+  std::stringstream tvstr;
+  tvstr << tv;
+  for (auto id : root) {
+    const DomainKey key(td, id);
+    auto it = pending_map_.find(key);
+    if (it == pending_map_.end()) {
+      // No consumer of this ID found. Just need to add this to
+      // eq_set_ but nothing equivalent with it yet
+      //std::cerr << "No consumer found: " << key << " of " <<
+      //tvstr.str() << std::endl;
+      continue;
+    }
+    if (mapAllConsumers(key)) {
```

Comment:
If found to be consistent, map consumer with producer, this is "recursive" meaning anything mapped to consumer previously (running this pass backwards) is mapped to this producer as well.